### PR TITLE
Fix the amounts in the Filter by Price block to reflect the shopper's selected currency.

### LINKED ISSUE #6024: Multicurrency - Filter by Price block doesn't convert the prices
### Describe the bug
When the user is switched to the geolocation currency (once the option for an auto switch is enabled), the Filter by Price still displays the values in default store currency.

### To Reproduce

1. Make sure to have plugins: WC Payments, WC Blocks installed
2. In WC Payments settings enable currency switch based on geolocation
3. Change the default store currency to different one than your local currency
4. Create a new post
5. Add Filter by Price and Products blocks
6. Save and go to frontend
7. Add a product to the cart
8. Refresh the page 

### Actual behavior
While all the products display the prices in your local currency and prices are converted, Filter by Price displays the correct currency, but value is not converted and still in store default currency.

### Expected behavior
Filter by Price should display min and max price converted to the geolocated currency.

### Example:
In this example: USD to PLN exchange is ~4.5. The max price of the store is 60 USD, so after conversion to PLN:

Expected: In PLN it should be 60 * 4.5 = 290 zł
Actual: Filter by Price still displays 60 zł (red rectangle)

image

Keep in mind there are two problems:
1. Currency will be picked up once the user session is established. That can be mimicked by adding a product to the store (allowing the filter to pick up the currency without the session is covered in this issue: https://github.com/Automattic/woocommerce-payments/issues/5996)
2. The scope of this particular issue: value is not converted even though

#### POC
This issue involves WC Payments and WC Blocks. Here's a POC showcasing a possible solution:
- On WCPay we should add a dedicated class in compatibility layer: https://github.com/Automattic/woocommerce-payments/pull/5984
- On WCBlocks we have to add the new filters: https://github.com/woocommerce/woocommerce-blocks/pull/8977

- Corresponding WCBlocks issue: https://github.com/woocommerce/woocommerce/issues/42407

#### Ref
More details available in per0F9-yg-p2

### PR DESCRIPTION
**Pull request by rafaelzaleski**
Fixes #6024 

#### Changes proposed in this Pull Request

- **Fixes the price ranges** displayed in the *Filter by Price* block to accurately reflect the shopper's selected currency.
- **Corrects the filter query amounts** to ensure the correct values are passed when filtering products in a currency different from the store's default currency.
- **Removes the applied min and max prices from the URL** when the shopper changes the selected currency, preventing incorrect filtering results.

#### Testing instructions

**Setup Requirements:**
- The *Filter by Price* block and the *Product Collection* block must be added to the same page for filtering to work correctly.
- Ensure WooPayments Multi-Currency is enabled, with the setting **"Automatically switch customers to their local currency if it has been enabled"** turned on.
- Use currencies with a large difference in amounts for testing, such as **JPY** and **USD**, to clearly observe the conversion differences.
- You can verify conversion rates in **WooPayments Multi-Currency Settings**.

---

#### Test Scenario 1: Currency Change Updates Min/Max Prices
1. Add the *Filter by Price* and *Product Collection* blocks to the same page.
2. Set the store's default currency to **USD**.
3. Change the currency to **JPY** using the currency switcher widget.
4. Verify that the min and max prices in the *Filter by Price* block are updated to reflect the correct converted values in **JPY**.

✅ **Expected Result:** The min and max prices reflect the correct conversion rates for **JPY**.

---

#### Test Scenario 2: Filter Products in the New Currency
1. Apply a min and max price range filter using the *Filter by Price* block in **USD**.
2. Change the currency to **JPY** using the currency switcher widget.
3. Set a new min and max price range in the *Filter by Price* block.
4. Verify that the products are filtered correctly based on the selected price range in **JPY**.

✅ **Expected Result:** Products are correctly filtered based on the price range in the new currency.

---

#### Test Scenario 3: URL Filter Removal on Currency Change
1. Apply a min and max price filter in **USD**.
2. Change the currency to **JPY** using the currency switcher widget.
3. Check the URL after changing the currency.

✅ **Expected Result:** The min and max prices are removed from the URL when the currency is changed.

---

#### Test Scenario 4: Reapply Filters After Currency Change
1. Change the currency to **JPY** using the currency switcher widget.
2. Apply a new min and max price range in the *Filter by Price* block.
3. Verify that the products are filtered correctly based on the new range in **JPY**.

✅ **Expected Result:** It is possible to apply new filters in the selected currency after the currency change.

---

#### Test Scenario 5: Geolocation-Based Currency Change
1. Set the store currency to a currency **different from the tester’s default local currency** (e.g., set the store currency to **USD** if the tester is in **Japan**).
2. Add the tester’s local currency (e.g., **JPY**) to the list of available currencies in **WooPayments Multi-Currency Settings**.
3. Open the page with the *Filter by Price* block in an **incognito tab**.
4. Verify that the currency automatically switches to the tester’s local currency based on their location.

✅ **Expected Result:** The currency switches to the tester’s local currency automatically, and the min and max prices in the *Filter by Price* block are updated with the correct converted values in that currency.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10059` or branch name `fix/6024-filter-by-price-does-not-convert-prices` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6024-filter-by-price-does-not-convert-prices) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 8e0d68fd7b54ba74d4f70d2d5883253bc686913d
- Build time: 2025-01-17 22:25:10 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/multi-currency/FrontendCurrencies.php [hunk 1]
```diff
@@ -137,6 +137,22 @@ public function selected_currency_changed() {
...
+	 *
+	 * Clears existing price filters when the currency is changed to prevent inconsistencies.
+	 *
+	 * @return void
+	 */
+	public function clear_url_price_params() {
+		if ( isset( $_GET['min_price'] ) || isset( $_GET['max_price'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+			$url = remove_query_arg( [ 'min_price', 'max_price' ] );
+
+			wp_safe_redirect( $url );
```
Comment (bborman22) [inline includes/multi-currency/FrontendCurrencies.php:151]: This causes a double page request when changing currency while a min/max price parameter is set. Is there any alternative approach we could use here to avoid that? Since it only happens in one unique case I don't think it's urgent to address, but wanted to ask still.

File: includes/multi-currency/FrontendPrices.php [hunk 2]
```diff
@@ -72,7 +72,134 @@ public function init_hooks() {
...
+	 * @param int       $page  The current page number.
+	 *
+	 * @return array The modified query variables.
+	 */
+	public function maybe_modify_price_ranges_query_var( $query, $block, $page ) {
+		if ( 'product' !== $query['post_type'] ) {
+			return $query;
+		}
+
+		if ( ! isset( $query['meta_query'] ) || ! is_array( $query['meta_query'] ) ) {
```
Comment (bborman22) [inline includes/multi-currency/FrontendPrices.php]: We can probably add a check for if meta_query is an empty array as well. I found that when stepping through and just changing the currency, I'd hit `convert_meta_query_price_filters` below, but with an empty array which properly exits early in the for loop, but we can probably skip altogether.

File: includes/multi-currency/FrontendPrices.php [hunk 3]
```diff
@@ -72,7 +72,134 @@ public function init_hooks() {
...
+					// Extract the relation and the nested queries.
+					$relation = $mq['relation'];
+
+					$modified_nested = $this->convert_meta_query_price_filters( $mq, $from_currency, $target_currency );
+
+					// Reconstruct the meta_query with the modified nested queries.
+					$mq = array_merge( [ 'relation' => $relation ], $modified_nested );
+				}
+			} elseif ( isset( $mq['key'] ) && '_price' === $mq['key'] && isset( $mq['value'] ) && is_numeric( $mq['value'] ) ) {
+					$converted_price = $this->multi_currency->get_raw_conversion( $mq['value'], $from_currency, $target_currency );
```
Comment (bborman22) [inline includes/multi-currency/FrontendPrices.php]: Nit: I think this is indented one too many, but the linter doesn't seem to have a problem with it, so might just be a convention I'm not familiar with?

File: includes/multi-currency/FrontendPrices.php [hunk 4]
```diff
@@ -72,7 +72,134 @@ public function init_hooks() {
...
+	 * from the 'from_currency' to the 'target_currency'.
+	 *
+	 * @param array  $meta_query       The meta_query array to traverse.
+	 * @param string $from_currency   The from currency code.
+	 * @param string $target_currency The target currency code.
+	 *
+	 * @return array The modified meta_query array.
+	 */
+	private function convert_meta_query_price_filters( $meta_query, $from_currency, $target_currency ) {
+		foreach ( $meta_query as &$mq ) {
```
Comment (bborman22) [inline includes/multi-currency/FrontendPrices.php:134]: Recursion always makes me a bit nervous on things like this, but I think this only runs into a problem if there was a circular reference, which I wouldn't expect to happen in this case. I also tested with additional filters being applied (in stock and ratings) and the elseif properly exited out when examining those so I think we're good. Just wanted to highlight if there were any concerns you had here that you'd want me to test or take a second look at?

**Review by bborman22 — CHANGES_REQUESTED** [review]
This tested really well and worked exactly as expected. I ran through all the test cases you listed with success and found that the price ranges and the filters all worked. I tested with USD, EUR, and JPY to get a range of prices and tested while switching between them all with success!

I also did some really rudimentary performance testing and didn't see any negative impact from running these conversions, oddly enough in one set of tests it actually was faster than the `develop` branch.

I left a couple notes, mostly just to double check my thinking. The only change request I'd have is to add a check for the empty array just to avoid an additional function call, but the others are just comments to get your opinion.

File: includes/multi-currency/FrontendCurrencies.php (hunk 1 above)
Comment (rafaelzaleski) [reply includes/multi-currency/FrontendCurrencies.php:151]: I tried removing the query parameters at runtime in PHP, but it didn’t work. 
I also considered removing these parameters when an option is selected in the currency switcher, but what happens if the merchant uses another method to switch the currency? That approach wouldn’t work with the geolocation feature either.

In the end, I believe the redirect is the cleaner and more reliable solution. Additionally, this redirect will occur in such a specific scenario that the impact of a page redirect is minimal, in my opinion. It only happens if you have pre-existing filters and then switch the currency (and you also need to have MCCY enabled). How often would that occur in typical user interactions?

File: includes/multi-currency/FrontendPrices.php (hunk 4 above)
Comment (rafaelzaleski) [reply includes/multi-currency/FrontendPrices.php:134]: As you mentioned, this would only be an issue with a circular reference in a malformed meta query array. Since we can’t guarantee that other plugins won’t cause such circular references, let’s be defensive and limit the recursion to 5 levels deep. This way, we avoid being the ones responsible for throwing a fatal error due to a malformed array.

File: includes/multi-currency/FrontendPrices.php (hunk 2 above)
Comment (rafaelzaleski) [reply includes/multi-currency/FrontendPrices.php]: Done in https://github.com/Automattic/woocommerce-payments/pull/10059/commits/436ad1e0b7e770480be33c36099cb6e77274761a

File: includes/multi-currency/FrontendPrices.php (hunk 3 above)
Comment (rafaelzaleski) [reply includes/multi-currency/FrontendPrices.php]: I don't know how this was not caught by the linter. 
Fixed in https://github.com/Automattic/woocommerce-payments/pull/10059/commits/8372689e4853b4b8569d1c65a9072c01ac1f9d05

File: includes/multi-currency/FrontendPrices.php (hunk 4 above)
Comment (rafaelzaleski) [reply includes/multi-currency/FrontendPrices.php:134]: Done in https://github.com/Automattic/woocommerce-payments/pull/10059/commits/8e0d68fd7b54ba74d4f70d2d5883253bc686913d

File: includes/multi-currency/FrontendCurrencies.php (hunk 1 above)
Comment (bborman22) [reply includes/multi-currency/FrontendCurrencies.php:151]: > Additionally, this redirect will occur in such a specific scenario that the impact of a page redirect is minimal, in my opinion.

Agreed! Just wanted to see if there were any alternatives, but like you said this seems the most reliable and simplest for a very small subset of use cases.

File: includes/multi-currency/FrontendPrices.php (hunk 3 above)
Comment (bborman22) [reply includes/multi-currency/FrontendPrices.php]: Thanks! Looks good!

File: includes/multi-currency/FrontendPrices.php (hunk 4 above)
Comment (bborman22) [reply includes/multi-currency/FrontendPrices.php:134]: Works great and at least adds a safety net for this function, looks good to me and thanks for adding this!

**Review by bborman22 — APPROVED** [review]
I confirmed all your latest updates addressed my concerns and the feature still worked as planned. I ran through another set of tests against it and all were successful!

LGTM!
